elements/lisk-utils/src/math/q.ts

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable no-bitwise */
+/*
+ * Copyright © 2022 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { MAX_FRAC, ONE, Q_OPERATION, TWO } from './constants';
+
+const numberToQ = (base: bigint, val: number): bigint => {
+	const denominator = TWO ** base;
+	if (Number.isInteger(val)) {
+		return BigInt(val) * denominator;
+	}
+	const [intStr, fractionalStr] = val.toString().split('.');
+	return (
+		BigInt(intStr) * denominator +
+		(BigInt(fractionalStr) * denominator) / BigInt(10) ** BigInt(fractionalStr.length)
+	);
+};
+
+export const q = (val: number | bigint | Buffer, base: number | bigint): Q =>
+	Q.fromValue(val, base);
+
+export class Q {
+	private readonly _base: bigint;
+	private readonly _val: bigint;
+
+	public constructor(val: bigint, base: number | bigint) {
+		this._base = BigInt(base);
+		if (this._base > MAX_FRAC) {
+			throw new Error('Base exceeds max fraction allowed.');
+		}
+		this._val = val;
+		this._base = BigInt(base);
+	}
+
+	public static fromValue(val: number | bigint | Buffer | Q, base: number | bigint) {
```

Comment:
Do we really need 1 big method that handles different input formats, where there is no overlapping logic at all?

I would prefer to break this down into 3 smaller methods:

* `fromNumber()`
* `fromBigInt()`
* `fromBuffer()`

Not sure that we need `fromQ()` at all, because we can directly do it using the constructor.

Reply:
in utils, if we accept different values, anyway it will require to check the type of the input. This is factory method so i think it's fine to receive different types.
the usage will be like
```
q96(3).add(q96(3))
```
instead of
```
Q.fromNumber(3, 96).add(Q.fromNumber(3, 96))
```